M.E.C.M. Add TryGetValue(ReadOnlySpan<char>) API

mgravell opened the pull request:
Implement `MemoryCache.TryGetValue` APIs taking `ReadOnlySpan<char>`, as approved in #110504

fix #110504

mgravell committed:
M.E.C.M. Add TryGetValue(ReadOnlySpan<char>) API

fix #110504

mgravell committed:
fix test glitch

mgravell commented on src/libraries/Microsoft.Extensions.Caching.Memory/tests/AltLookupTests.cs:
```diff
@@ -0,0 +1,78 @@
+            // not found after removing
+            cache.Remove(stringKey);
+            cache.Set(stringKey, cachedValue);
+
```
fixed in 28e7ad049c4fd5db0c55c4d10d4a8421dce39fe5

stephentoub commented on src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs:
```diff
@@ -211,9 +211,74 @@ public bool TryGetValue(object key, out object? result)
-            if (coherentState.TryGetValue(key, out CacheEntry? tmp))
+            if (!coherentState.TryGetValue(key, out CacheEntry? entry))
+            {
+                entry = null;
```
Is entry not already null when TryGetValue returns false?

mgravell replied:
4f601494c54bc080224f8b2bb540a004ae0da751

stephentoub commented on src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs:
```diff
@@ -671,11 +736,31 @@ private sealed class CoherentState
+#if NET9_0_OR_GREATER
+            internal bool TryGetValue(ReadOnlySpan<char> key, [NotNullWhen(true)] out CacheEntry? entry)
+                => _useStringAltLookup ? _stringAltLookup.TryGetValue(key, out entry)
+                    : _stringEntries.TryGetValue(new string(key), out entry); // <== we do not expect this path to be hit; chaos fallback only
```
Why add code for a case that can never happen? Shouldn't this just always use _stringAltLookup?

mgravell replied:
It was meant defensively, but: if this API is reliable (in this context), fine: I'll remove the extra field and the switches

mgravell replied:
4f601494c54bc080224f8b2bb540a004ae0da751

stephentoub commented on src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs:
```diff
@@ -671,11 +736,31 @@ private sealed class CoherentState
+
+            public CoherentState()
+            {
+                _useStringAltLookup = _stringEntries.TryGetAlternateLookup<ReadOnlySpan<char>>(out _stringAltLookup);
```
Can this just use GetAlternateLookup instead?

mgravell replied:
see 4f601494c54bc080224f8b2bb540a004ae0da751

mgravell commented on src/libraries/Microsoft.Extensions.Caching.Memory/ref/Microsoft.Extensions.Caching.Memory.cs:
```diff
@@ -37,6 +40,13 @@ protected virtual void Dispose(bool disposing) { }
+
+#if NET9_0_OR_GREATER
+        [OverloadResolutionPriority(1)]
+        public bool TryGetValue(ReadOnlySpan<char> key, out object? result) { throw null; }
```
Documentation is in src - let me know if I'm wrong (humans., I mean)

mgravell commented on src/libraries/Microsoft.Extensions.Caching.Memory/ref/Microsoft.Extensions.Caching.Memory.cs:
```diff
@@ -37,6 +40,13 @@ protected virtual void Dispose(bool disposing) { }
+        [OverloadResolutionPriority(1)]
+        public bool TryGetValue(ReadOnlySpan<char> key, out object? result) { throw null; }
+        [OverloadResolutionPriority(1)]
+        public bool TryGetValue<TItem>(ReadOnlySpan<char> key, out TItem? value) { throw null; }
```
Documentation is in src - let me know if I'm wrong (humans., I mean)

mgravell committed:
PR feedback; be less defensive:

- demand GetAlternateLookup works on the target runtimes
- rely on documented "out set to default on failure" lookup behavior

mgravell commented:
@stephentoub made less defensive in 4f601494c54bc080224f8b2bb540a004ae0da751, per feedback

mgravell commented on src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs:
```diff
@@ -671,11 +736,31 @@ private sealed class CoherentState
             private readonly ConcurrentDictionary<object, CacheEntry> _nonStringEntries = new ConcurrentDictionary<object, CacheEntry>();
+
+#if NET9_0_OR_GREATER
+            private readonly bool _useStringAltLookup;
```
4f601494c54bc080224f8b2bb540a004ae0da751

mgravell committed:
Merge branch 'main' into marc/MECMSpanGet

vcsjones commented on src/libraries/Microsoft.Extensions.Caching.Memory/ref/Microsoft.Extensions.Caching.Memory.cs:
```diff
@@ -37,6 +40,13 @@ protected virtual void Dispose(bool disposing) { }
+
+#if NET9_0_OR_GREATER
+        [OverloadResolutionPriority(1)]
+        public bool TryGetValue(ReadOnlySpan<char> key, out object? result) { throw null; }
```
The approved API calls the second parameter `value`, not `result`.

The mismatch of the parameter name between `ref` and `src` is also why the build failed.

mgravell replied:
Will fix, thanks. I caught it in one place, not the other.

mgravell replied:
474041ddfce80ea9eb98c2937a618f70720b2e9f

vcsjones commented on src/libraries/Microsoft.Extensions.Caching.Memory/ref/Microsoft.Extensions.Caching.Memory.cs:
```diff
@@ -4,6 +4,9 @@
 // ------------------------------------------------------------------------------
 
+using System;
+using System.Runtime.CompilerServices;
```
Generally the `ref` source does not have `using`s. Every type is qualified with its own namespace.

mgravell replied:
474041ddfce80ea9eb98c2937a618f70720b2e9f

mgravell committed:
address PR feedback

mgravell committed:
Merge branch 'marc/MECMSpanGet' of https://github.com/mgravell/runtime into marc/MECMSpanGet

stephentoub commented on src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs:
```diff
@@ -211,9 +211,68 @@ public bool TryGetValue(object key, out object? result)
             DateTime utcNow = UtcNow;
```
Does this need to be done before the TryGetValue call, or could it be consolidated into your PostProcessTryGetValue helper?

mgravell replied:
It is a valid question; my primary aim here was to retain the existing behaviour, but I think it is fair to challenge it - I suspect the only *critical* thing is that `utcNow` doesn't change between the 4 places where it is used (and as a secondary aim: that we don't pay the cost of fetching it 4 times), but as long as we *do* check expiration, I can't think of any way in which it is less "valid" if we defer the time lookup to after the initial fetch. I'll try moving it.

stephentoub commented on src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs:
```diff
@@ -211,9 +211,68 @@ public bool TryGetValue(object key, out object? result)
+        /// <param name="value">The located value or null.</param>
+        /// <returns>True if the key was found.</returns>
+        /// <remarks>This method allows values with <see cref="string"/> keys to be queried by content without allocating a new <see cref="string"/> instance.</remarks>
+        [OverloadResolutionPriority(1)]
```
@333fred, @jaredpar, just FYI on more use of ORPA.

mgravell committed:
consolidate "now" lookup into PostProcessTryGetValue

stephentoub commented:
This didn't actually run CI

mgravell commented:
Well shoot, it was all briefly green. Will fix: sorry.

On Thu, 20 Feb 2025, 16:00 Stephen Toub, ***@***.***> wrote:

> It also merged with
> https://github.com/dotnet/runtime/pull/112695/files#r1963779247
> unaddressed
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/dotnet/runtime/pull/112695#issuecomment-2671934466>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAEHMEQJOGMXH7NN7MLLBT2QX3Z7AVCNFSM6AAAAABXOPZ4Q6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMNZRHEZTINBWGY>
> .
> You are receiving this because you modified the open/close state.Message
> ID: ***@***.***>
> [image: stephentoub]*stephentoub* left a comment (dotnet/runtime#112695)
> <https://github.com/dotnet/runtime/pull/112695#issuecomment-2671934466>
>
> It also merged with
> https://github.com/dotnet/runtime/pull/112695/files#r1963779247
> unaddressed
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/dotnet/runtime/pull/112695#issuecomment-2671934466>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAEHMEQJOGMXH7NN7MLLBT2QX3Z7AVCNFSM6AAAAABXOPZ4Q6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMNZRHEZTINBWGY>
> .
> You are receiving this because you modified the open/close state.Message
> ID: ***@***.***>
>
